usr/src/uts/common/io/scsi/adapters/smartpqi/smartpqi_main.c

```diff
@@ -0,0 +1,419 @@
+		if (s->s_rescan != NULL) {
+			(void) untimeout(s->s_rescan);
+			s->s_rescan = NULL;
+		}
+
+		if (s->s_watchdog != 0) {
+			(void) untimeout(s->s_watchdog);
+			s->s_watchdog = 0;
+		}
+
+		if (s->s_error_dma != NULL) {
+			pqi_free_single(s, s->s_error_dma);
+			s->s_error_dma = NULL;
+		}
+		if (s->s_adminq_dma != NULL) {
+			pqi_free_single(s, s->s_adminq_dma);
+			s->s_adminq_dma = NULL;
+		}
+		if (s->s_queue_dma != NULL) {
+			pqi_free_single(s, s->s_queue_dma);
+			s->s_queue_dma = NULL;
+		}
+
+		/* ---- Safe to always call ---- */
+		pqi_free_io_resource(s);
+
+		if (s->s_cmd_cache != NULL) {
+			kmem_cache_destroy(s->s_cmd_cache);
+			s->s_cmd_cache = NULL;
+		}
+
+		if (s->s_events_taskq != NULL) {
+			ddi_taskq_destroy(s->s_events_taskq);
+			s->s_events_taskq = NULL;
+		}
+		if (s->s_complete_taskq != NULL) {
+			ddi_taskq_destroy(s->s_complete_taskq);
+			s->s_complete_taskq = NULL;
+		}
+
+		while ((devp = list_head(&s->s_devnodes)) != NULL) {
+			/* ---- Better not be any active commands ---- */
+			ASSERT(list_is_empty(&devp->pd_cmd_list));
+
+			ddi_devid_free_guid(devp->pd_guid);
+			if (devp->pd_pip != NULL)
+				(void) mdi_pi_free(devp->pd_pip, 0);
+			if (devp->pd_pip_offlined)
+				(void) mdi_pi_free(devp->pd_pip_offlined, 0);
+			list_destroy(&devp->pd_cmd_list);
+			mutex_destroy(&devp->pd_mutex);
+			list_remove(&s->s_devnodes, devp);
+			PQI_FREE(devp, sizeof (*devp));
+		}
+		list_destroy(&s->s_devnodes);
+		mutex_destroy(&s->s_mutex);
+		mutex_destroy(&s->s_io_mutex);
+		mutex_destroy(&s->s_intr_mutex);
+
+		cv_destroy(&s->s_quiescedvar);
+		smartpqi_unregister_hba(s);
+		smartpqi_unregister_intrs(s);
+
+		if (s->s_mem_timeo != 0) {
+			mutex_enter(&s->s_mem_mutex);
+			(void) untimeout(s->s_mem_timeo);
+			s->s_mem_timeo = 0;
+			mutex_exit(&s->s_mem_mutex);
+			mutex_destroy(&s->s_mem_mutex);
+		}
+
+		if (s->s_time_of_day != 0) {
+			(void) untimeout(s->s_time_of_day);
+			s->s_time_of_day = 0;
+		}
+
+		ddi_soft_state_free(pqi_state, instance);
+		ddi_prop_remove_all(dip);
+	}
+
+	return (DDI_SUCCESS);
+}
+
+/*ARGSUSED*/
+static int
+smartpqi_power(dev_info_t *dip, int component, int level)
+{
+	/* We don't register any power components yet. */
+	return (DDI_SUCCESS);
+}
+
+static int
+smartpqi_quiesce(dev_info_t *dip)
+{
+	pqi_state_t	s;
+	int		instance;
+
+	/*
+	 * ddi_get_soft_state is lock-free, so is safe to call from
+	 * quiesce.  Furthermore, pqi_hba_reset uses only the safe
+	 * drv_usecwait() and register accesses.
+	 */
+	instance = ddi_get_instance(dip);
+	if ((s = ddi_get_soft_state(pqi_state, instance)) != NULL) {
+		if (pqi_hba_reset(s)) {
+			return (DDI_SUCCESS);
+		}
+	}
+	/* If we couldn't quiesce for any reason, play it safe and reboot. */
+	return (DDI_FAILURE);
+}
+
+/*ARGSUSED*/
+static int
+smartpqi_ioctl(dev_t dev, int cmd, intptr_t data, int mode, cred_t *credp,
+    int *rval)
+{
+	/* Arguably we could just use nodev for the entry point. */
```

Comment:
So why don't you?

Reply:
The above comment was added by me -- as part of my other contribs, and I agree, we should just remove it.